### ISSUE TITLE
I want to rewrite hosting with provision4 and drupal 8

### ISSUE THREAD

**Issue by ipumpkin** [provision-ops/provision#9]
Hello @jonpugh,  thanks for your provision4 project, that's a really amazing work. 
I want to rewrite a front end with provision4 and Drupal 8, can you give me some opinion or instruction. Have you has some plan of that? Thanks a lot.

**Comment by jonpugh** [comment]
Oh, that's great to hear! Thank you for the kind words.

I have ideas but no solid plan just yet. I think the best route is to create a separate set of modules called provision_ui. This module can require provision/cli directly, so we can get metadata about context types there.

If we include provision4 in the drupal module, perhaps we can override config_path to be in the Drupal sites folder (www-data readable). That way we can call `provision save` directly from drupal (not in the task queue).

Then, we can point provision4 CLI config path to that sites/x/files/provision folder, effectively joining the front and back ends.

**Comment by jonpugh** [comment]
As for the components, we'd need form fields for all of the properties. I'd prefer not to use field API but try to save and load from YML directly.

Perhaps we create provision4 contexts as Drupal 8 Config entities?

**Comment by jonpugh** [comment]
I actually already toyed around with upgrading devshop to do this for 2.x: See https://cgit.drupalcode.org/devmaster/tree/composer.json?h=8.x-2.x

I wouldn't continue this branch, but it gives you an idea.

Again I think provision_ui module should be a separate module, then something like devshop would require it and add on enhanced features

**Comment by ipumpkin** [comment]
Thank you very much for your idea. I will give it a try first.

**Comment by ipumpkin** [comment]
I have tried to run provision command in Drupal, but the command  always return error code. There is some code:
`public static function commandSave(EntityInterface $context = null, $new = false){
    $class = Context::getClassName($context->getEntityTypeId());
    $method = "option_documentation";
    $options = [];
    foreach ($class::{$method}() as $key => $property) {
      if(!empty($context->{$key}->value)){
        $options[$key] = $context->{$key}->value;
      }
    }
    $options['context_type'] = $context->getEntityTypeId();
    $options['context_class'] = $class;
    return static::provisionCommand('save',$context,$options);
  } 
public static function provisionCommand($command, EntityInterface $context = null, $options = []){
    
    $logger = \Drupal::logger('Provision UI');
    
    $argv = [$command];
    if($context){
      $argv[] = "@{$context->getEntityTypeId()}_{$context->machine_name->value}";
    }
    foreach ($options as $key => $value){
      $argv[] = "--{$key}={$value}";
    }
    $argv[] = "-a";
    $input = new ArgvInput($argv);
    $output = new BufferedOutput();
    $io = new ProvisionStyle($input, $output);
    
    $config = new Config($io,FALSE);

    // Create the app.
    $provision = new \Aegir\Provision\Provision($config, $input, $output);
    
    // Run the app.
    $status_code = $provision->run($input, $output);
    
    $result = [
      'status_code' => $status_code,
      'output' => $output->fetch()
    ];
    
   
    $logger->debug("Command:" . print_r($command,1));
    $logger->debug("Argv:" . print_r($argv,1));
    $logger->debug("Result:" . print_r($result,1));
    return $result;
  } `


When run commandSave() ,i will get error info below. Can you give me some suggestions?


`
Result:Array ( [status_code] => 1 [output] => The "--context_class" option does not exist. list [--raw] [--format FORMAT] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-c|--context [CONTEXT]] [--] [] [] )

Argv:Array ( [0] => save [1] => @server_test_44 [2] => --context_class=\Aegir\Provision\Context\ServerContext [3] => --remote_host=localhost [4] => --script_user=aegir [5] => --aegir_root=\Users\ipump [6] => --context_type=server [7] => -a )
`

**Comment by ipumpkin** [comment]
I got it : the first element of the array is the name of the running application.
`$argv = [Provision::APPLICATION_NAME, $command];`

**Comment by jonpugh** [comment]
@ipumpkin We shouldn't run the commands from Drupal, we should run $context->save() directly, or push this into an actual "service" Drupal 8 could use...

I'm still a novice at the Drupal8 services architecture, but perhaps others like @cweagans can chime in on how we can start to create a Provision service that can be leveraged by both this CLI and Drupal8.

**Comment by cweagans** [comment]
I don't know a lot about Provision 4, but if I was going to sit down and architect something like this, I'd probably use the core Queue API to keep track of the operations that need to run, and then `drush queue-run provision` periodically in the background as a user that has permission to make whatever modifications Provision needs to make.

That'd be pretty similar to earlier versions of Aegir, but we'd be offloading the queue handling to core APIs, which makes it a) pretty robust and b) easily swappable by end users (for instance, if a user wanted to use AWS SQS instead of the default Drupal database queue, it would be pretty trivial for them to make that change without patching anything from Aegir).

**Comment by jonpugh** [comment]
Thanks @cweagans that is good advice, but what I'm talking about here is how we might include provision classes into Drupal 8 so we can access all that stuff directly and write the context yml files.

Queue stuff is quite a ways down the road.

**Comment by cweagans** [comment]
Ah, I see. Okay, can you point me toward the class that you'd want to be able to use in Drupal?

**Comment by jonpugh** [comment]
I'm thinking almost the entire thing can be lifted out into a service package and become a dependency of the CLI and the Drupal module. 

All the Context ServerContext SiteContext and Services classes could be included in a module, then have form code that maps to the context and service options.

It's not much different from provision 3.x, just composerified and namespaced.

**Comment by cweagans** [comment]
Alright, so if you wanted to include ServerContext, you need to look at the constructor and work up to the top of the dependency chain. I see that the ServerContext constructor takes a Provision object as an arg to the constructor. Looking at the Provision class, I see that it needs a Config, InputInterface, and OutputInterface object. Looking at Config, it needs a ProvisionStyle object, which needs InputInterface and OutputInterface.

I'd probably start by implementing an InputInterface and OutputInterface compatible service that makes sense inside of Drupal (maybe your Input/Output goes to a db log or something). Then you can take those implementations and wire them up as dependencies to the ProvisionStyle service. Once you have ProvisionStyle, you can wire it up as a dependency to Config and so on.

Doing it that way is a pretty substantial amount of work.

In this case, I'd probably still recommend using a queue + queue runner to write out the context. Setting up separate queues for stuff is basically free in Drupal, so you can have a `context_save` queue that just takes data and passes it to whatever CLI command is appropriate. Later, it's easy to add another queue that keeps track of applying changes to infrastructure.

**Comment by ipumpkin** [comment]
There is  a module i has made: [provision_ui](https://github.com/ipumpkin/provision_ui).Using eck entity map to provision context and task, config entity map to provision service, Use core queue to run the tasks.
Can run the base provison command.

**Comment by jonpugh** [comment]
Thank you @ipumpkin that is awesome.

I am restarting the efforts on this project, but I'm focused as quick and dirty an upgrade as I can for Hosting 7.x-4.x

The provision_ui module you created is great, and I see things going there eventually. 

First focus is rewrite of the CLI, then stabilize the whole 4.x suite, then rewrite of the drupal modules.  

We'll get there!

**Comment by colans** [comment]
Instead of duplicating work, why not use the Drupal 8 front-end we've already started working on for [Aegir 5](https://gitlab.com/aegir/aegir/)?

The back-end is meant to be pluggable so while we've started with Ansible (and hope to also support Kubernetes), there's no reason why this couldn't work with Provision 4 as well.

Aegir-specific entities are already defined, and the queue runner is [Celery](https://en.wikipedia.org/wiki/Celery_(software)).  More information on the architecture is available at https://gitlab.com/aegir/aegir/wikis/architecture.
